File: GroceryMobileApp/GroceryApp/APIModels.swift
import Foundation

// MARK: - Auth

struct LoginRequest: Encodable {
    let email: String
    let password: String
}

struct RegisterRequest: Encodable {
    let firstName: String
    let lastName: String
    let email: String
    let password: String
    let phoneNumber: String?
}

struct AuthResponse: Decodable {
    let success: Bool
    let token: String?
    let expiresAt: Date?
    let user: UserDTO?
    let errors: [String]?
}

struct UserDTO: Codable, Identifiable {
    let id: UUID
    let firstName: String
    let lastName: String
    let email: String
    let phoneNumber: String?
    let profileImageUrl: String?
    let roles: [String]?

    var fullName: String { "\(firstName) \(lastName)" }
}

// MARK: - Categories

struct CategoryDTO: Decodable, Identifiable {
    let id: UUID
    let name: String
    let description: String?
    let imageUrl: String?
    let isActive: Bool
    let productCount: Int

    /// Map to the app's GroceryCategory
    var asGroceryCategory: GroceryCategory {
        GroceryCategory(id: id, name: name, emoji: emojiForCategory(name))
    }
}

private func emojiForCategory(_ name: String) -> String {
    let map: [String: String] = [
        "fruits": "🍎", "vegetables": "🥕", "veggie": "🥕",
        "meats": "🥩", "meat": "🥩", "snacks": "🍿",
        "drinks": "🥤", "beverages": "🥤", "dairy": "🧀",
        "bakery": "🍞", "bread": "🍞", "seafood": "🦐",
        "frozen": "🧊", "organic": "🌿", "condiments": "🧂",
    ]
    return map[name.lowercased()] ?? "🛒"
}

// MARK: - Products

struct ProductDTO: Decodable, Identifiable {
    let id: UUID
    let name: String
    let description: String?
    let price: Decimal
    let discountPrice: Decimal?
    let stockQuantity: Int
    let unit: String?
    let isActive: Bool
    let categoryId: UUID
    let categoryName: String
    let categories: [ProductCategoryDTO]?
    let images: [ProductImageDTO]
    let averageRating: Double
    let reviewCount: Int
    let createdAt: Date?

    /// Map to the app's GroceryProduct
    var asGroceryProduct: GroceryProduct {
        let priceDouble = NSDecimalNumber(decimal: price).doubleValue
        let originalDouble = discountPrice != nil ? priceDouble : nil
        let discountDouble = discountPrice.map { NSDecimalNumber(decimal: $0).doubleValue }
        let actualPrice = discountDouble ?? priceDouble
        let discountText: String? = if let orig = originalDouble, orig > 0 {
            "\(Int(((orig - actualPrice) / orig) * 100))% off"
        } else {
            nil
        }
        // Prefer fullUrl for image display, fall back to imageUrl
        let primaryImage = images.first(where: { $0.isPrimary }) ?? images.first
        let imageURL = primaryImage?.fullUrl ?? primaryImage?.imageUrl

        // Debug: print image URL resolution
        print("🖼️ [\(name)] images: \(images.count), primary: \(primaryImage?.imageUrl ?? "none"), fullUrl: \(primaryImage?.fullUrl ?? "nil"), resolved: \(imageURL ?? "nil")")

        return GroceryProduct(
            id: id,
            name: name,
            location: categoryName,
            price: discountDouble ?? priceDouble,
            originalPrice: discountPrice != nil ? priceDouble : nil,
            discount: discountText,
            emoji: emojiForCategory(categoryName),
            category: categoryName,
            imageURL: imageURL
        )
    }

    /// Additional category names (excluding the primary categoryName)
    var additionalCategoryNames: [String] {
        guard let categories else { return [] }
        return categories.map(\.name).filter { $0 != categoryName }
    }

    /// All image full URLs for gallery display
    var imageGalleryURLs: [String] {
        images.sorted(by: { $0.sortOrder < $1.sortOrder })
            .compactMap { $0.fullUrl ?? $0.imageUrl }
    }
}

struct ProductCategoryDTO: Decodable, Identifiable {
    let id: UUID
    let name: String
}

struct ProductImageDTO: Decodable, Identifiable {
    let id: UUID
    let imageUrl: String
    let fullUrl: String?
    let isPrimary: Bool
    let sortOrder: Int

    /// Best URL to use for display — prefers fullUrl from API
    var displayUrl: String {
        fullUrl ?? imageUrl
    }
}

// MARK: - Cart

struct CartItemDTO: Decodable, Identifiable {
    let id: UUID
    let productId: UUID
    let productName: String
    let productImageUrl: String?
    let unitPrice: Decimal
    let quantity: Int
    let totalPrice: Decimal
}

struct AddToCartRequest: Encodable {
    let productId: UUID
    let quantity: Int
}

struct UpdateCartItemRequest: Encodable {
    let quantity: Int
}

// MARK: - Orders

struct OrderDTO: Decodable, Identifiable {
    let id: UUID
    let orderNumber: String
    let subTotal: Decimal
    let discountAmount: Decimal
    let deliveryFee: Decimal
    let totalAmount: Decimal
    let status: String
    let notes: String?
    let createdAt: Date
    let items: [OrderItemDTO]?
    let payment: PaymentSummaryDTO?
    let address: OrderAddressDTO?
    let statusHistory: [OrderStatusHistoryDTO]?

    var asOrderItem: OrderItem {
        let formatter = DateFormatter()
        formatter.dateFormat = "MMM d, yyyy"
        return OrderItem(
            id: id,
            orderNumber: orderNumber,
            date: formatter.string(from: createdAt),
            items: items?.reduce(0) { $0 + $1.quantity } ?? 0,
            total: NSDecimalNumber(decimal: totalAmount).doubleValue,
            status: OrderStatus.from(status),
            orderRemarks: notes ?? "",
            paymentMethod: payment?.method ?? "Unknown",
            paymentDetail: ""
        )
    }
}

struct OrderItemDTO: Decodable {
    let productId: UUID
    let productName: String
    let unitPrice: Decimal
    let quantity: Int
    let totalPrice: Decimal
}

struct PaymentSummaryDTO: Decodable {
    let method: String
    let status: String
    let paidAt: Date?
}

struct OrderAddressDTO: Decodable {
    let label: String
    let street: String
    let city: String
    let province: String
    let zipCode: String

    var fullAddress: String {
        [street, city, province, zipCode].filter { !$0.isEmpty }.joined(separator: ", ")
    }
}

struct OrderStatusHistoryDTO: Decodable, Identifiable {
    let status: String
    let notes: String?
    let createdBy: String
    let createdAt: Date

    var id: String { "\(status)-\(createdAt.timeIntervalSince1970)" }
}

struct CreateOrderRequest: Encodable {
    let addressId: UUID?
    let voucherCode: String?
    let notes: String?
}

// MARK: - Addresses

struct AddressDTO: Decodable, Identifiable {
    let id: UUID
    let label: String
    let street: String
    let city: String
    let province: String
    let zipCode: String
    let country: String?
    let deliveryInstructions: String?
    let contactNumber: String?
    let latitude: Double?
    let longitude: Double?
    let isDefault: Bool

    var fullAddress: String {
        [street, city, province, zipCode].filter { !$0.isEmpty }.joined(separator: ", ")
    }

    var asAddressItem: AddressItem {
        AddressItem(
            id: id,
            label: label,
            address: fullAddress,
            isDefault: isDefault,
            deliveryInstructions: deliveryInstructions ?? "",
            contactNumber: contactNumber ?? "",
            latitude: latitude ?? 0,
            longitude: longitude ?? 0
        )
    }
}

struct CreateAddressRequest: Encodable {
    let label: String
    let street: String
    let city: String
    let province: String
    let zipCode: String
    let country: String?
    let deliveryInstructions: String?
    let contactNumber: String?
    let latitude: Double?
    let longitude: Double?
    let isDefault: Bool
}

struct UpdateAddressRequest: Encodable {
    let label: String?
    let street: String?
    let city: String?
    let province: String?
    let zipCode: String?
    let country: String?
    let deliveryInstructions: String?
    let contactNumber: String?
    let latitude: Double?
    let longitude: Double?
    let isDefault: Bool?
}

// MARK: - Favorites

struct FavoriteDTO: Decodable, Identifiable {
    let id: UUID?
    let productId: UUID
    let productName: String?
    let price: Decimal?
    let discountPrice: Decimal?
    let imageUrl: String?
    let addedAt: Date?
    let product: ProductDTO?

    /// Build a GroceryProduct from the flat favorite data or nested product
    var asGroceryProduct: GroceryProduct? {
        if let product { return product.asGroceryProduct }
        guard let productName else { return nil }
        let priceDouble = price.map { NSDecimalNumber(decimal: $0).doubleValue } ?? 0
        let discountDouble = discountPrice.map { NSDecimalNumber(decimal: $0).doubleValue }
        return GroceryProduct(
            id: productId,
            name: productName,
            location: "",
            price: discountDouble ?? priceDouble,
            originalPrice: discountPrice != nil ? priceDouble : nil,
            discount: nil,
            emoji: "🛒",
            category: "",
            imageURL: imageUrl
        )
    }
}

// MARK: - Payment Methods

struct PaymentMethodDTO: Decodable, Identifiable {
    let id: UUID
    let name: String
    let detail: String?
    let paymentType: String
    let icon: String?
    let isDefault: Bool
    let createdAt: Date?
}

struct CreatePaymentMethodRequest: Encodable {
    let name: String
    let detail: String?
    let paymentType: String
    let icon: String?
    let isDefault: Bool
}

struct UpdatePaymentMethodRequest: Encodable {
    let name: String?
    let detail: String?
    let paymentType: String?
    let icon: String?
    let isDefault: Bool?
}

// MARK: - Vouchers

struct VoucherDTO: Decodable, Identifiable {
    let id: UUID
    let code: String
    let description: String?
    let type: String
    let value: Decimal
    let maxDiscount: Decimal?
    let minimumSpend: Decimal
    let usageLimit: Int
    let usedCount: Int
    let isActive: Bool
    let startDate: Date?
    let expiryDate: Date

    var asVoucherItem: VoucherItem {
        let formatter = DateFormatter()
        formatter.dateFormat = "MMM d, yyyy"
        let discountText: String = if type.lowercased() == "percentage" {
            "\(NSDecimalNumber(decimal: value).intValue)%"
        } else {
            "$\(NSDecimalNumber(decimal: value).intValue)"
        }
        return VoucherItem(
            id: id,
            code: code,
            description: description ?? "",
            discount: discountText,
            minOrder: minimumSpend > 0 ? "Min. $\(NSDecimalNumber(decimal: minimumSpend).intValue)" : "No minimum",
            validUntil: formatter.string(from: expiryDate),
            isActive: isActive && expiryDate > Date()
        )
    }
}

struct ApplyVoucherRequest: Encodable {
    let code: String
    let cartTotal: Decimal
}

struct VoucherValidationResult: Decodable {
    let isValid: Bool
    let errorMessage: String?
    let discountAmount: Decimal
    let voucher: VoucherDTO?
}

// MARK: - Reviews

struct ReviewDTO: Decodable, Identifiable {
    let id: UUID
    let userId: UUID
    let userName: String
    let productId: UUID
    let rating: Int
    let comment: String?
    let createdAt: Date
    let photos: [ReviewPhotoDTO]?
}

struct ReviewPhotoDTO: Decodable, Identifiable {
    let id: UUID
    let photoUrl: String
    let sortOrder: Int
}

struct CreateReviewRequest: Encodable {
    let productId: UUID
    let orderId: UUID
    let rating: Int
    let comment: String?
    let photoUrls: [String]?
}

// MARK: - Payments

struct CheckoutPaymentRequest: Encodable {
    let orderId: UUID
    let method: Int // 0=Card, 1=ApplePay, 2=GCash, 3=PayMaya, 4=COD
    let stripeToken: String?
    let returnUrl: String?
}

struct PaymentResultDTO: Decodable {
    let success: Bool
    let paymentId: UUID
    let status: String
    let redirectUrl: String?
    let failureReason: String?
}

// MARK: - Notifications

struct NotificationDTO: Decodable, Identifiable {
    let id: UUID
    let title: String
    let message: String
    let type: String?
    let referenceId: String?
    let isRead: Bool
    let createdAt: Date
}

// MARK: - User Settings

struct UserSettingDTO: Decodable {
    let key: String
    let value: String
}

struct UpdateUserSettingRequest: Encodable {
    let key: String
    let value: String
}

File: GroceryWeb/src/GroceryAndroid/app/src/main/java/com/sanshare/groceryapp/data/remote/ApiModels.kt
package com.sanshare.groceryapp.data.remote

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import java.net.URI

// ── Auth ──────────────────────────────────────────────────────────────────────

@Serializable
data class LoginRequest(
    val email: String,
    val password: String,
    val deviceGuid: String? = null,
    val osVersion: String? = null,
    val hardwareVersion: String? = null,
)

@Serializable
data class RegisterRequest(
    val firstName: String,
    val lastName: String,
    val email: String,
    val password: String,
    val phoneNumber: String? = null,
    val deviceGuid: String? = null,
    val osVersion: String? = null,
    val hardwareVersion: String? = null,
)

@Serializable
data class AuthResponse(
    val success: Boolean,
    val token: String? = null,
    val expiresAt: String? = null,
    val user: UserDto? = null,
    val errors: List<String>? = null,
    val requiresEmailVerification: Boolean? = null,
)

@Serializable
data class VerifyEmailRequest(val email: String, val code: String)

@Serializable
data class VerifyEmailResponse(
    val success: Boolean,
    val token: String? = null,
    val error: String? = null,
)

@Serializable
data class VerifyPhoneRequest(val code: String)

@Serializable
data class VerifyPhoneResponse(val success: Boolean, val error: String? = null)

@Serializable
class EmptyRequest

@Serializable
data class SimpleMessageResponse(val message: String? = null)

@Serializable
data class UserDto(
    val id: String,
    val firstName: String,
    val lastName: String,
    val email: String,
    val phoneNumber: String? = null,
    val profileImageUrl: String? = null,
    val roles: List<String>? = null,
    val isEmailVerified: Boolean? = null,
    val isPhoneVerified: Boolean? = null,
) {
    val fullName: String get() = "$firstName $lastName".trim().ifEmpty { email }
}

// ── Categories ────────────────────────────────────────────────────────────────

@Serializable
data class CategoryDto(
    val id: String,
    val name: String,
    val description: String? = null,
    val imageUrl: String? = null,
    val emoji: String? = null,
    val isActive: Boolean = true,
    val productCount: Int = 0,
)

// ── Products ──────────────────────────────────────────────────────────────────

@Serializable
data class ProductDto(
    val id: String,
    val name: String,
    val description: String? = null,
    val price: Double,
    val discountPrice: Double? = null,
    val stockQuantity: Int = 0,
    val unit: String? = null,
    val isActive: Boolean = true,
    val categoryId: String,
    val categoryName: String,
    val categories: List<ProductCategoryDto>? = null,
    val images: List<ProductImageDto> = emptyList(),
    val averageRating: Double = 0.0,
    val reviewCount: Int = 0,
    val createdAt: String? = null,
) {
    val displayPrice: Double get() = discountPrice ?: price
    val primaryImageUrl: String? get() = resolveApiUrl(
        images.firstOrNull { it.isPrimary }?.displayUrl
            ?: images.firstOrNull()?.displayUrl
    )
    val discountPercent: Int? get() {
        if (discountPrice == null || discountPrice >= price) return null
        return ((price - discountPrice) / price * 100).toInt()
    }
}

@Serializable
data class ProductCategoryDto(val id: String, val name: String)

@Serializable
data class ProductImageDto(
    val id: String,
    val imageUrl: String,
    val fullUrl: String? = null,
    val isPrimary: Boolean = false,
    val sortOrder: Int = 0,
    val dateCreated: String? = null,
    val dateModified: String? = null,
) {
    val displayUrl: String get() = resolveApiUrl(fullUrl ?: imageUrl) ?: imageUrl
}

private fun resolveApiUrl(url: String?): String? {
    if (url.isNullOrBlank()) return null
    if (url.startsWith("http", ignoreCase = true)) return rebuildUsingBaseUrl(url)
    val normalizedBase = ApiConfig.BASE_URL.trimEnd('/')
    val normalizedPath = if (url.startsWith("/")) url else "/$url"
    return "$normalizedBase$normalizedPath"
}

private fun rebuildUsingBaseUrl(url: String): String {
    return try {
        val source = URI(url)
        val path = source.path.orEmpty()
        val normalizedPath = when {
            path.contains("/uploads/products/", ignoreCase = true) ->
                path.substringAfter("/uploads/products/").substringBefore("?")
            path.contains("/products/", ignoreCase = true) ->
                path.substringAfter("/products/").substringBefore("?")
            else -> path.substringAfterLast('/').substringBefore("?")
        }.trimStart('/')

        if (normalizedPath.isBlank()) url
        else "${ApiConfig.BASE_URL.trimEnd('/')}/uploads/products/$normalizedPath"
    } catch (_: Exception) {
        url
    }
}

@Serializable
data class PagedResult<T>(
    val items: List<T>,
    val totalCount: Int,
    val page: Int,
    val pageSize: Int,
    val totalPages: Int,
)

// ── Cart ──────────────────────────────────────────────────────────────────────

@Serializable
data class CartItemDto(
    val id: String,
    val productId: String,
    val productName: String,
    val productImageUrl: String? = null,
    val productImageFullUrl: String? = null,
    val productImageDateModified: String? = null,
    val unitPrice: Double,
    val quantity: Int,
    val totalPrice: Double,
    val remarks: String? = null,
)

@Serializable
data class AddToCartRequest(
    val productId: String,
    val quantity: Int,
    val remarks: String? = null,
)

@Serializable
data class UpdateCartItemRequest(
    val quantity: Int,
    val remarks: String? = null,
)

// ── Orders ────────────────────────────────────────────────────────────────────

@Serializable
data class OrderDto(
    val id: String,
    val orderNumber: String,
    val subTotal: Double,
    val discountAmount: Double = 0.0,
    val deliveryFee: Double = 0.0,
    val platformFee: Double? = null,
    val otherCharges: Double? = null,
    val totalAmount: Double,
    val status: String,
    val notes: String? = null,
    val voucherCode: String? = null,
    val createdAt: String,
    val deliveryDate: String? = null,
    val deliveryTimeSlot: String? = null,
    val riderId: String? = null,
    val riderName: String? = null,
    val riderContact: String? = null,
    val riderImageUrl: String? = null,
    val actualDeliveryDate: String? = null,
    val items: List<OrderItemDto>? = null,
    val payment: PaymentSummaryDto? = null,
    val address: OrderAddressDto? = null,
    val statusHistory: List<OrderStatusHistoryDto>? = null,
    val reviews: List<OrderReviewDto>? = null,
)

@Serializable
data class OrderItemDto(
    val productId: String,
    val productName: String,
    val productImageUrl: String? = null,
    val productImageDateModified: String? = null,
    val unitPrice: Double,
    val quantity: Int,
    val totalPrice: Double,
    val remarks: String? = null,
)

@Serializable
data class PaymentSummaryDto(
    val method: String,
    val status: String,
    val paidAt: String? = null,
)

@Serializable
data class OrderAddressDto(
    val label: String,
    val street: String,
    val city: String,
    val province: String,
    val zipCode: String,
    val deliveryInstructions: String? = null,
    val contactNumber: String? = null,
) {
    val fullAddress: String get() = listOf(street, city, province, zipCode)
        .filter { it.isNotBlank() }.joinToString(", ")
}

@Serializable
data class OrderStatusHistoryDto(
    val status: String,
    val notes: String? = null,
    val createdBy: String,
    val createdAt: String,
)

@Serializable
data class OrderReviewDto(
    val id: String,
    val userName: String,
    val productName: String,
    val rating: Int,
    val comment: String? = null,
    val createdAt: String,
    val photos: List<OrderReviewPhotoDto>? = null,
)

@Serializable
data class OrderReviewPhotoDto(
    val id: String,
    val photoUrl: String,
    val sortOrder: Int = 0,
)

@Serializable
data class CreateOrderRequest(
    val addressId: String? = null,
    val voucherCode: String? = null,
    val notes: String? = null,
    val deliveryDate: String? = null,   // "yyyy-MM-dd" plain string
    val deliveryTimeSlot: String? = null,
    val platformFee: Double = 2.0,
    val otherCharges: Double = 1.0,
)

// ── Addresses ─────────────────────────────────────────────────────────────────

@Serializable
data class AddressDto(
    val id: String,
    val label: String,
    val street: String,
    val city: String,
    val province: String,
    val zipCode: String,
    val country: String? = null,
    val deliveryInstructions: String? = null,
    val contactNumber: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val isDefault: Boolean = false,
) {
    val fullAddress: String get() = listOf(street, city, province, zipCode)
        .filter { it.isNotBlank() }.joinToString(", ")
}

@Serializable
data class CreateAddressRequest(
    val label: String,
    val street: String,
    val city: String,
    val province: String,
    val zipCode: String,
    val country: String? = "Philippines",
    val deliveryInstructions: String? = null,
    val contactNumber: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val isDefault: Boolean = false,
)

@Serializable
data class UpdateAddressRequest(
    val label: String? = null,
    val street: String? = null,
    val city: String? = null,
    val province: String? = null,
    val zipCode: String? = null,
    val country: String? = null,
    val deliveryInstructions: String? = null,
    val contactNumber: String? = null,
    val latitude: Double? = null,
    val longitude: Double? = null,
    val isDefault: Boolean? = null,
)

// ── Favorites ─────────────────────────────────────────────────────────────────

@Serializable
data class FavoriteDto(
    val id: String? = null,
    val productId: String,
    val productName: String? = null,
    val price: Double? = null,
    val discountPrice: Double? = null,
    val imageUrl: String? = null,
    val addedAt: String? = null,
    val product: ProductDto? = null,
)

// ── Payment Methods ───────────────────────────────────────────────────────────

@Serializable
data class PaymentMethodDto(
    val id: String,
    val name: String,
    val detail: String? = null,
    val paymentType: String,
    val icon: String? = null,
    val isDefault: Boolean = false,
    val createdAt: String? = null,
)

@Serializable
data class CreatePaymentMethodRequest(
    val name: String,
    val detail: String? = null,
    val paymentType: String,
    val icon: String? = null,
    val isDefault: Boolean = false,
)

// ── Vouchers ──────────────────────────────────────────────────────────────────

@Serializable
data class VoucherDto(
    val id: String,
    val code: String,
    val description: String? = null,
    val type: String,
    val value: Double,
    val maxDiscount: Double? = null,
    val minimumSpend: Double,
    val usageLimit: Int,
    val usedCount: Int,
    val isActive: Boolean,
    val startDate: String? = null,
    val expiryDate: String,
)

@Serializable
data class ApplyVoucherRequest(val code: String, val cartTotal: Double)

@Serializable
data class VoucherValidationResult(
    val isValid: Boolean,
    val errorMessage: String? = null,
    val discountAmount: Double = 0.0,
    val voucher: VoucherDto? = null,
)

// ── Payments ──────────────────────────────────────────────────────────────────

@Serializable
data class CheckoutPaymentRequest(
    val orderId: String,
    val method: Int,   // 4=COD
    val stripeToken: String? = null,
    val returnUrl: String? = null,
)

@Serializable
data class PaymentResultDto(
    val success: Boolean,
    val paymentId: String,
    val status: String,
    val redirectUrl: String? = null,
    val failureReason: String? = null,
)

// ── Special Offers ────────────────────────────────────────────────────────────

@Serializable
data class SpecialOfferDto(
    val id: String,
    val categoryId: String? = null,
    val title: String,
    val subtitle: String,
    val emoji: String,
    val imageUrl: String? = null,
    val backgroundColorHex: String,
    val sortOrder: Int = 0,
    val isActive: Boolean = true,
)

@Serializable
data class TodayDealDto(
    val id: String,
    val productId: String,
    val sortOrder: Int = 0,
    val isActive: Boolean = true,
    val product: ProductDto,
)

// ── Notifications ─────────────────────────────────────────────────────────────

@Serializable
data class NotificationDto(
    val id: String,
    val title: String,
    val message: String,
    val type: String? = null,
    val referenceId: String? = null,
    val isRead: Boolean = false,
    val createdAt: String,
)

// ── Create Review ─────────────────────────────────────────────────────────────

@Serializable
data class CreateReviewRequest(
    val productId: String,
    val orderId: String,
    val rating: Int,
    val comment: String? = null,
    val photoUrls: List<String>? = null,
)

@Serializable
data class UploadReviewPhotosResponse(
    val urls: List<String>,
)
